### ISSUE TITLE
support unique port for consul / event-driven service discovery

### ISSUE THREAD

**Issue by chen23** [F5Networks/f5-appsvcs-extension#117]
##### ISSUE TYPE

 - Feature Idea

Currently AS3 expects that you are using a fixed port number for service discovery when using Consul or event-driven service discovery.  When using consul with a service like registrator, https://gliderlabs.github.io/registrator/latest/, it is possible to have pool members that have unique port numbers (i.e. 192.168.128.11:8081, 192.168.128.11:8082, etc...).  Ideally it should be possible to support this use case using AS3.

**Comment by focrensh** [comment]
I can break this into 2 requests:

- Consul AT-576 Internally: For this are you looking to have it poll the "services" endpoint on consul vs the nodes endpoint? 

- Event-Drive: AT-575 Internally

**Comment by chen23** [comment]
Yes, I believe that hitting the Service would make sense, but I have to admit its been a while since I looked at this in any detail.  The following is a mock-up this idea: https://github.com/f5devcentral/f5-demo-consul

**Comment by chen23** [comment]
I've create a PR for adding support for f5-cloud-libs-consul to support this use case: https://github.com/F5Networks/f5-cloud-libs-consul/pull/3

It also requires modifying f5-service-discovery getNodes.js with the following change (this uses memberAddressType as a method to toggle between the change in behavior.  Ideally a new parameter should be added to enable this to prevent breaking existing behavior):
```
diff --git a/src/lib/getNodes.js b/src/lib/getNodes.js
index 3745422..ce0c17d 100644
--- a/src/lib/getNodes.js
+++ b/src/lib/getNodes.js
@@ -106,6 +106,13 @@ module.exports = {

 function getNode(owner, memberAddressType) {
     const ip = memberAddressType.toLowerCase() === 'public' ? owner.ip.public : owner.ip.private;
+    if (owner.port && ip && memberAddressType.toLowerCase() === 'public') {
+        return {
+            ip,
+            id: `${owner.id}-${memberAddressType.toLowerCase()}`,
+            port: owner.port
+        };
+    }
     if (ip) {
         return {
             ip,
```

**Comment by jmcalalang** [comment]
Another customer request for ports to be harvested from consul. Contact Grey Schober or for details.

**Comment by dstokesf5** [comment]
AUTOTOOL-575 was resolved in release 3.15.0. 
AUTOTOOL-576 is still open.

**Comment by chen23** [comment]
This is now possible in AS3 3.24 using a custom jmespathquery. 

https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/declarations/discovery.html#using-service-discovery-to-find-consul-ports
